### PR TITLE
[compatible] Stop txn/network after slot feature

### PR DESCRIPTION
**Pull request by joaosreis**
TL;DR implements RFC #14138

This PR introduces two new optional compile-time configuration parameters, `slot_tx_end` and `slot_chain_end`. These define, respectively, the slot at which the node will stop producing/validating non-empty blocks (defined as being blocks that produce an empty staged ledger diff), and the slop at which the node stops producing/accepting any blocks at all. This is a feature part of the hard-fork mechanism.

Summary of changes:

- Add optional `slot-tx-end` parameter to CLI.
- BP always uses an empty set of transactions to produce blocks after the `slot-tx-end` slot instead of fetching them from the transaction pool.
- The RPC function for sending user commands rejects commands after the `slot-tx-end` slot.
- The node doesn't try to add incoming transactions to the network pool after the `slot-tx-end` slot.
- The transaction pool drops all transactions when the `slot-tx-end` slot is reached.

Some unit tests (specifically, for the transaction pool) were added. The complete set of expected behaviors was demonstrated in a sandbox network.

Merge PRs to other branches:

- #14770
- #14773
- #15000

### CODE REVIEW

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

File: src/app/test_executive/slot_end_test.ml [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  (* TODO: find a way to avoid this type alias (first class module signatures restrictions make this tricky) *)
+  type network = Network.t
+
+  type node = Network.Node.t
+
+  type dsl = Dsl.t
+
+  let num_extra_keys = 100
+
+  let slot_tx_end = Mina_compile_config.slot_tx_end
```
Comment (nholland94) [inline src/app/test_executive/slot_end_test.ml:20]: This is conflating the compile-time-config of the test executive with the compile-time-config of the daemon we are running. These are not one in the same, and won't necessarily match.

Comment (nholland94) [reply src/app/test_executive/slot_end_test.ml:20]: To deal with this, as specified in the comment section of the RFC, these configuration options should be exposed in the node configuration so that the test executive can configure them directly for the test. Then we don't need to pair together various compile time configurations to different test cases.

File: src/app/test_executive/slot_end_test.ml [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      wait_for t
+        (Wait_condition.nodes_to_initialize (String.Map.data all_nodes))
+    in
+    let%bind () =
+      section_hard "wait for 3 blocks to be produced (warm-up)"
+        (wait_for t (Wait_condition.blocks_to_be_produced 3))
+    in
+    let end_t =
+      Time.add (Time.now ())
+        (Time.Span.of_ms @@ float_of_int (num_slots * window_ms))
```
Comment (nholland94) [inline src/app/test_executive/slot_end_test.ml:123]: This isn't necessarily aligned to the network's notion of slots. You probably want to compute an offset from the genesis timestamp.

**Review by nholland94 — CHANGES_REQUESTED** [review]

File: src/lib/mina_graphql/mina_graphql.ml [hunk 3]
```diff
@@ -3065,20 +3065,22 @@ module Mutations = struct
...
-        | Ok user_command ->
-            Ok
-              { Types.UserCommand.With_status.data = user_command
-              ; status = Unknown
-              }
-        | Error e ->
-            Error ("Couldn't send user_command: " ^ Error.to_string_hum e) )
-    | `Bootstrapping ->
-        return (Error "Daemon is bootstrapping")
+    match Mina_compile_config.slot_tx_end with
```
[the remark this answers is not in the document]
Comment (joaosreis) [reply src/lib/mina_graphql/mina_graphql.ml:3068]: In one of the first iterations of the corresponding RFC, it was discussed changing the transaction pool logic, but it was decided to keep it unchanged.

As for rejecting txns at the CLI level, it was suggested to do it at the GraphQL layer. I checked the code, and calling the CLI to send payments uses GraphQL, so, inherently, txns are rejected when using CLI.

On an initial implementation that changed the txn pool, txns were being rejected, either if they came from CLI/GraphQL or the gossip network. In this current implementation, we have no mechanism to deal with txns coming from the gossip network. Regardless, rejecting txns is not necessary to produce empty blocks, it was added simply as an improvement to UX.

Please let me know if you need me to address this in another way.

File: src/lib/block_producer/block_producer.ml [hunk 4]
```diff
@@ -111,207 +111,257 @@ end
...
-              Block_time.now time_controller
-              |> Block_time.to_span_since_epoch |> Block_time.Span.to_ms
-            in
-            O1trace.sync_thread "generate_consensus_transition" (fun () ->
-                Consensus_state_hooks.generate_transition
-                  ~previous_protocol_state ~blockchain_state ~current_time
-                  ~block_data ~supercharge_coinbase
-                  ~snarked_ledger_hash:previous_ledger_hash ~genesis_ledger_hash
-                  ~supply_increase ~logger ~constraint_constants ) )
+      Option.iter (Option.bind Mina_compile_config.slot_chain_end ~f:slot_diff)
```
[the remark this answers is not in the document]
Comment (joaosreis) [reply src/lib/block_producer/block_producer.ml:138]: I moved the logging logic to the outer `run` BP function.

File: src/app/test_executive/slot_end_test.ml (hunk 1 above)
Comment (joaosreis) [reply src/app/test_executive/slot_end_test.ml:20]: Since overriding these settings in runtime was not added (as per [this discussion](https://github.com/MinaProtocol/mina/pull/14138#discussion_r1370378912)), what would be the best way to do this? I was thinking of adding these parameters to the runtime configuration, but hiding the override compile time config logic behind a feature flag, enabling it only on the build profile used for the integration test (`devnet`?)
Also, Lucy has to be extended to support these new runtime config params, right?

**Comment by joaosreis** [comment]
!ci-build-me

File: src/lib/mina_graphql/mina_graphql.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (nholland94) [reply src/lib/mina_graphql/mina_graphql.ml:3068]: The CLI doesn't always use graphql. For instance, sending batches of payments does not use graphql, it uses the rpc interface https://github.com/MinaProtocol/mina/blob/develop/src/app/cli/src/init/client.ml#L433.

I would still urge you to move this error into the transaction pool so that we have full coverage of all the places where a user can submit commands.

File: src/app/test_executive/slot_end_test.ml (hunk 1 above)
Comment (nholland94) [reply src/app/test_executive/slot_end_test.ml:20]: Apologies for the confusion from that comment. In the comment you linked, I meant to say that it should not be overridable via the CLI. In another discussion we had mentioned that it should be overridable via the runtime config only. Exposing it via runtime config doesn't have the same risk that exposing it via the CLI does. Most users never edit the runtime config, since it will usually make your node unable to join the chain, and we don't document the runtime config options. https://github.com/MinaProtocol/mina/pull/14138#discussion_r1329357416

Lucy does not require additional changes if you expose this via runtime config. The runtime config is set by every test in Lucy. This option should automatically be exposed, and if it isn't, it's a 2 line change to expose it, since Lucy just takes the test config you give and generates the runtime config from it.

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
!ci-build-me

File: src/app/test_executive/slot_end_test.ml (hunk 1 above)
Comment (joaosreis) [reply src/app/test_executive/slot_end_test.ml]: Added these two parameters to the runtime configuration with a mechanism to retrieve them if set or default to the compile-time values if not. Also made some small changes to Lucy to use these new config parameters.

**Comment by georgeee** [comment]
> There will be a configuration parameter set at compile-time that will define the slot at which the node will stop processing transactions.
The previous configuration cannot be overridable at runtime, by design, as this compromises the safety of the daemon software.

Does this excerpt from RFC mean that there should be no option to set the two parameters via CLI flag?

CC @mrmr1993 @nholland94

**Comment by nholland94** [comment]
@georgeee Yes, there should not be a CLI flag. We did align in the RFC discussion to allow a runtime configuration override for integration testing purposes, which can be left undocumented.

File: src/app/test_executive/slot_end_test.ml [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+                 "non-empty block after slot_tx_end. block slot since genesis: \
+                  %s, txn count: %d, snark work count: %d, coinbase: %d"
+                 (Mina_numbers.Global_slot.to_string block.slot_since_genesis)
+                 block.command_transaction_count block.snark_work_count
+                 block.coinbase
+             in
+             ok_if_true msg
+               ( Mina_numbers.Global_slot.(
+                   of_uint32 block.slot_since_genesis < slot_tx_end)
+               || block.command_transaction_count = 0
```
Comment (georgeee) [inline src/app/test_executive/slot_end_test.ml:187]: NIT: I'd put conjunction group into braces for clarity (not necessary, but required me to remember operator precedence to be sure it works)

File: src/lib/network_pool/indexed_pool.ml [hunk 6]
```diff
@@ -351,6 +355,8 @@ let current_global_slot t =
   | None ->
       current_slot
 
+(* let slot_tx_end t = t.slot_tx_end *)
```
Comment (georgeee) [inline src/lib/network_pool/indexed_pool.ml:358]: Remove comment

File: src/lib/network_pool/indexed_pool.ml [hunk 7]
```diff
@@ -725,114 +731,307 @@ let rec add_from_gossip_exn :
...
+    Consensus.Data.Consensus_time.(
+      to_global_slot
+        (of_time_exn ~constants:consensus_constants
+           (Block_time.now time_controller) ))
   in
-  (* C4 *)
-  match Map.find t.all_by_sender fee_payer with
-  | None ->
-      (* nothing queued for this sender *)
+  match slot_tx_end with
```
Comment (georgeee) [inline src/lib/network_pool/indexed_pool.ml:748]: I suggest to flatten this code a little bit.

Something like the following would fit nicely:

```
let%bind () = 
Result.ok_if_true ~error:After_slot_tx_end @@
Option.value_map ~f:Global_slot.(fun slot_tx_end' ->
current_global_slot >= slot_tx_end') ~default:true slot_tx_end in
```

Comment (georgeee) [reply src/lib/network_pool/indexed_pool.ml:748]: TODO for George: re-review the code of the function after flattening (now changes are inflated by increase of layerness)

File: src/lib/network_pool/indexed_pool.ml [hunk 8]
```diff
@@ -725,114 +731,307 @@ let rec add_from_gossip_exn :
...
+       ; _
+       } as t ) cmd ->
+  let open Result.Let_syntax in
+  let current_global_slot =
+    Consensus.Data.Consensus_time.(
+      to_global_slot
+        (of_time_exn ~constants:consensus_constants
+           (Block_time.now time_controller) ))
+  in
+  match slot_tx_end with
```
Comment (georgeee) [inline src/lib/network_pool/indexed_pool.ml:1014]: Please flatten this code as suggested above

File: src/lib/network_pool/transaction_pool.ml [hunk 9]
```diff
@@ -133,18 +178,30 @@ module Diff_versioned = struct
...
+          "This transaction was submitted after the slot defined to stop \
+           accepting transactions"
   end
 
   module Rejected = struct
     [%%versioned
     module Stable = struct
       [@@@no_toplevel_latest_type]
 
+      module V2 = struct
```
Comment (georgeee) [inline src/lib/network_pool/transaction_pool.ml:191]: Note for future: discuss on how this affects berkeley versioning types (whould we bump some types' versions there too?)

File: src/lib/network_pool/transaction_pool.ml [hunk 10]
```diff
@@ -2379,4 +2450,40 @@ let%test_module _ =
...
+            Test.Resource_pool.Diff.unsafe_apply pool
+            @@ Envelope.Incoming.local
+                 (extract_signed_commands independent_cmds)
+          in
+          [%test_eq: pool_apply] (Ok independent_cmds')
+            (accepted_commands apply_res) ;
+          assert_pool_txs independent_cmds' ;
+          Deferred.unit )
+
+    let%test_unit "transactions added after slot_tx_end are rejected" =
```
Comment (georgeee) [inline src/lib/network_pool/transaction_pool.ml:2474]:
```suggestion
    let%test_unit "transactions added at slot_tx_end are rejected" =
```

Comment (georgeee) [reply src/lib/network_pool/transaction_pool.ml:2474]: Should we add a test case for "after"?

**Review by georgeee — CHANGES_REQUESTED** [review]

File: src/app/test_executive/slot_end_test.ml (hunk 5 above)
Comment (joaosreis) [reply src/app/test_executive/slot_end_test.ml:187]: I certainly agree. It just turns out the formatter doesn't let me keep them :)

File: src/lib/network_pool/indexed_pool.ml (hunk 7 above)
Comment (joaosreis) [reply src/lib/network_pool/indexed_pool.ml:748]: Good suggestion. In this case, the slot comparison condition needs to be inverted. Before it was the condition where it would fail, now it should be the condition where it should proceed.

File: src/lib/network_pool/transaction_pool.ml (hunk 9 above)
Comment (joaosreis) [reply src/lib/network_pool/transaction_pool.ml:191]: This is the first time I'm updating versioned types, so I'm unsure on how this would affect 
`berkeley` types.
Regardless, this PR has merge conflicts with `berkeley` and `develop`, so we need to merge this feature into there as well. We can then upgrade these types there. Does this solve the issue?

Comment (georgeee) [reply src/lib/network_pool/transaction_pool.ml:191]: Yes, for sure. This is the way to proceed. My main point was that if we add a new numbered type here (`V2`) which existed in different form in `berkeley`, we probably should make sure to use `V3` in `berkeley`. I.e. if types between `compatible` and `berkeley` differ, `berkeley` should be +1 from `compatible`. That said, if you will find an existing `V3` in `berkeley` for a type that is `V2` in `compatible`, it's safe to change this type right away without introducing a new number.

Comment (georgeee) [reply src/lib/network_pool/transaction_pool.ml:191]: `berkeley` and `develop` are expected to have the same types, so once you remake this change against `berkeley`, there should be no problem with preparing a PR against `develop`.

File: src/lib/network_pool/transaction_pool.ml (hunk 10 above)
Comment (joaosreis) [reply src/lib/network_pool/transaction_pool.ml]: It surely doesn't hurt.

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by joaosreis** [comment]
@georgeee please give a look at the changes I just pushed.

**Review by georgeee — APPROVED** [review]

**Comment by joaosreis** [comment]
!ci-build-me

**Comment by georgeee** [comment]
!ci-build-me

**Review by ymekuria — APPROVED** [review]

**Comment by joaosreis** [comment]
!ci-build-me
